[Merged by Bors] - Silence annoying rustfmt config warnings

CleanCut opened the pull request:
Silence those [annoying rustfmt config warnings](https://github.com/bevyengine/bevy/pull/1499/checks?check_run_id=1950282111#step:5:66) that happen because we have unstable rustfmt options in `rustfmt.toml`, but we run it in stable on CI.  Thanks to @Ratysz for [calling it out](https://github.com/bevyengine/bevy/pull/1499#issuecomment-783190586). 😄 

The final approach we settled on was to comment out the unstable options in `rustfmt.toml`.  Those who are using `nightly` may  uncomment the unstable options locally if they wish. Once the options stabilize, we can uncomment them again.

We also decided that instead of fixing the alias, we would remove the alias entirely so that we do not introduce a custom `.cargo/config.toml` that would conflict with users' custom version of the same file. This means that instead of using a `cargo ci` alias you should use `cargo run -p ci` or `cargo run --package ci` instead.

<details><summary>Original Approach (abandoned)</summary>
<p>

_We decided **not** to go this way..._

In my quest to find a portable way to filter out the warnings I switched the library used to execute commands from `xshell` to `duct` (as advised by the `xshell` project itself when you want to do less simple things).  This still uses the "xtask" pattern of using a cargo command alias and a rust project for what would have usually been done with a bash script (on posix), just a different helper library is being used internally.

NOTE 1: Also, thanks to some sleuthing by @DJMcNab we were able to fix the broken cargo alias.  The issue turned out to be that `.cargo/config.toml` was being ignored because of `.gitignore`.

NOTE 2: This is a [known breaking change](https://github.com/bevyengine/bevy/pull/1309#discussion_r564023753) for anyone working on bevy who has their own local `.cargo/config.toml`.
</p>
</details>

NiklasEi commented on tools/ci/src/main.rs:
```diff
@@ -1,19 +1,47 @@
+        .stderr
+        .lines()
+        .filter_map(|l| l.ok())
+        .filter(|l| !l.starts_with("Warning"))
```
Could we make this a bit more specific so that we don't silence all warnings, but just the intended ones due to not running nightly?

CleanCut replied:
I'd prefer to stick with filtering by `Warning`.

My concern with more specific filtering is that we may end up with a brittle setup where we have to track the various specific config settings being warned about [as things change](https://github.com/bevyengine/bevy/pull/1479) multiplied across the different Rust versions our users use.

I looked through the `rustfmt` source code and found that `Warning` only occurs on the sorts of config errors we don't care about. I also see that this is the approach they use themselves [in rustfmt's own CI](https://github.com/rust-lang/rustfmt/blob/257311c304fddf578d8d4041ddf129533390b8ef/ci/integration.sh#L54).

A better long-term solution would be to teach `rustfmt` itself a `--no-config-warnings` option.  Or maybe figure out a way to only emit the warning _exactly once_.  Either would be more elegant than dealing with it in the output like we've had to do.

cart replied:
I think we should probably just remove the unstable commands from config.toml _and_ the filter here. That way we are maximally safe. We could just comment out the unstable commands for easier access from people (like me) that want the "better" formatting they provide. That feels safer and is makes the implementation easier to grok.

CleanCut replied:
✅ Unstable commands removed from `rustfmt.toml`
✅ Filtering commit reverted.

DJMcNab commented:
I'd rather if this just didn't check the formatting at all on ci with a non stable toolchain

Then always ran `cargo +nightly fmt --all` when run locally.

MinerSebas commented on tools/ci/src/main.rs:
```diff
@@ -1,19 +1,47 @@
+use duct::cmd;
 
 fn main() {
     // When run locally, results may from actual CI runs triggered by .github/workflows/ci.yml
```
```suggestion
    // When run locally, results may differ from actual CI runs triggered by .github/workflows/ci.yml
```

MinerSebas commented on tools/ci/src/main.rs:
```diff
@@ -1,19 +1,47 @@
 
     // See if any code needs to be formatted
-    cmd!("cargo fmt --all -- --check")
+    println!("$ cargo fmt --all -- --check");
+    let errput = cmd!("cargo", "fmt", "--all", "--", "--check")
```
Would it be possible, to only use the check flag, when this is run as Part of CI?
It would be more ergonomic, for Users if they don't have to fix it manually.

CleanCut replied:
I don't want to do that _in this PR._ A design goal of mine is that the command `cargo ci` _by itself_ act as identically as possible across all platforms and versions of Rust.

I'm certainly open to the ideas of either adding another alias (`cargo ci-local`?) or an argument (`cargo ci --auto`?) to get that changed behavior.  First, I'd like to see that someone actually wants it (maybe you?).  Then I think it would be best to implement that in it's own pull request. If that's something you feel like taking on, I'd be happy to be your reviewer and/or collaborator. 👍🏼

CleanCut commented:
> I'd rather if this just didn't check the formatting at all on ci with a non stable toolchain

@DJMcNab Your wish has already been granted!  In CI this already only runs on stable Rust.  Here's the current config on `main`:

https://github.com/bevyengine/bevy/blob/bc4fe9b186db3f32eef38fc4241289df480fa949/.github/workflows/ci.yml#L36-L39

> Then always ran `cargo +nightly fmt --all` when run locally.

Opinion noted. This is really up to the user, unfortunately. We don't place requirements on whether the user uses stable or nightly.  This means that a local user may see different results than CI if they're using a different version of Rust than CI's current stable.  Unfortunately, I think that's just the reality of how things need to be at the moment.  I don't see a way to avoid that unless we constrain what version of Rust a user selects.  The topic of which Rust to use was explored starting around here: https://github.com/bevyengine/bevy/pull/1309#issuecomment-766556798 

We _could_ add some documentation (a la #1507) stating that you should use the latest stable rust _iff_ you want `cargo ci` to exactly match CI -- but that may be confusing when contrasted with suggestions to use nightly in other places for best compile times.

CleanCut commented:
It occurred to me that `.cargo/config.toml` was in `.gitignore` so that we don't stomp users' changes.  So, I restored that entry in `.gitignore`.  We'll just need to be careful to add `-f` or `--force` to git commands when we, ourselves, modify that file. For example, `git add -f .cargo/config.toml`.

DJMcNab commented:
> @DJMcNab Your wish has already been granted!  In CI this already only runs on stable Rust. 

Except 'stable Rust' is exactly the opposite of 'not stable'

I'm saying we shouldn't be having these warnings, because it's significantly easier just to only run it in ci on nightly

Then we can actually check that the code follows the rustfmt config we have set up

And also mean `cargo xtask fmt` could be an alias for `cargo +nightly fmt`

DJMcNab commented:
Will force pushing allow our changes to be pulled?

I don't have a good enough git intuition to determine what will happen here

CleanCut commented:
@DJMcNab Thank you for clarifying your thoughts. Regarding nightly vs stable, the decision to go ahead with stable was settled on recently in https://github.com/bevyengine/bevy/pull/1309#issuecomment-766556798, so I don't feel like it needs to be rehashed here.

Regarding your Git question, _forcing_ is only necessary to add your changes to the index before you commit. No forcing will be necessary during a push, nor will pulling be affected.  In other words, special care only needs to be taken by someone who changes the `.cargo/config.toml` to make sure that change is added to a local commit if they want the changes to be included in their pull request -- after that step (adding a change to the index to be committed) nothing else is affected by `.gitignore`.  This is already how things were (and what I failed to notice when I created the `.cargo/config.toml` file earlier).

DJMcNab commented:
I'd seen that but hadn't read it fully - in that case I'd argue we just shouldn't have a `rustfmt.toml` checked in at the root, and instead have it somewhere else.  Then `cargo xtask fmt` would be `cargo +nightly fmt -- config-path <path_to_rusfmt_toml>`

And the change to ignore warnings in this PR would not be needed

bjorn3 commented on .cargo/config.toml:
```diff
@@ -0,0 +1,2 @@
+[alias]
+ci = "run --package ci"
```
nit: missing trailing newline

CleanCut replied:
If trailing newlines are important, we should add a check to CI. 

> I'd seen that but hadn't read it fully - in that case I'd argue we just shouldn't have a `rustfmt.toml` checked in at the root, and instead have it somewhere else. Then `cargo xtask fmt` would be `cargo +nightly fmt -- config-path <path_to_rusfmt_toml>`

AFAIK this would break @cart's desired experience. He uses nightly, and wants those options to take affect in his environment.

bjorn3 replied:
In git's view the newline is part of the current line and as such adding a missing newline to insert a line below the final line will cause the old final line to be considered as changed.

CleanCut replied:
Thanks for pointing out, that's good to know!  I encourage you to look for a way to integrate that check into one of the linters' configs if it bothers you.

In general, if you use the `Insert a suggestion` feature for nit-level concerns I'd be happy to hit the "accept change" button to commit it with you as co-author. Otherwise, I typically don't worry about nits unless I've got to modify the code in question for some other reason anyway.  I just wanted to make sure that you knew that I wasn't trying to ignore you or disagree with you.

CleanCut replied:
I was looking at this PR again tonight, so I went ahead and added that newline. 😄 

By the way, I'm baffled as to why there wasn't a trailing newline in the first place. Take a look at my "Even Better Toml" config:

<img width="505" alt="image" src="https://user-images.githubusercontent.com/5838512/109409028-5c6c2a00-794c-11eb-8110-c0eb8326643d.png">

I went and checked to see if [super-linter](https://github.com/github/super-linter) (which we're using for markdown linting) has a toml linter that I could add to CI, and it doesn't.  😞  Guess I'll just keep a sharper eye out for this happening in the future.

bjorn3 commented on .cargo/config.toml:
```diff
@@ -0,0 +1,2 @@
+[alias]
```
I currently have `.cargo/config.toml` symlinked to `.cargo/config_fast_builds`. This change will prevent doing so.

CleanCut replied:
@cart requested this approach [knowing it would cause this sort of conflict](https://github.com/bevyengine/bevy/pull/1309#discussion_r564023753).  That is to say, I'm acknowledging this problem as an accepted downside of this approach.

CleanCut replied:
Oh!  Though I should mention that since `.cargo/config.toml` is ignored by Git by default when looking at changes to include in commits, this should be straightforward to work around:

1. Delete your symlink
2. Now you can pull or switch to a branch that includes these changes
3. Then you can locally make whatever changes to `.cargo/config.toml` and it will be ignored.  Either re-establish your symlink, or copy the contents of `config_fast_builds` into it, or whatever you want.

bjorn3 replied:
The disadvantage of that is that changes to config_fast_build are no longer automatically applied. In addition git rebase complains about changed files, thus requiring git stash and not forgetting git stash pop.

CleanCut replied:
Ya, that sounds annoying. 😦

DJMcNab replied:
One solution should be to make a `.cargo/config.toml` symlinked to `config_fast_builds` but a level higher in the tree.

CleanCut replied:
@DJMcNab TIL! Nice find! [The docs say](https://doc.rust-lang.org/cargo/reference/config.html#hierarchical-structure) they'll merge `.cargo/config.toml` from _all_ parent directories.  So if I'm understanding correctly, if your `bevy` repo is located at `projects/bevy`, you can add a symlink:

* `projects/.cargo/config.toml -> ../bevy/.cargo/config_fast_builds`

...and the contents of that would be merged into whatever is in `bevy/.cargo/config.toml`.  It looks like the drawback of this approach is that any other rust project under `projects` will also be affected.

DJMcNab replied:
Yeah, that's right

My project structure is:
```
repositories/
| priroda/
| <...>
| bevy/
 | .cargo/config.toml
 | my_bevy_game/
 | assistance/ (cargo workspace with one target directory for when I need to clone someone else's project)
 | bevy_mod_picking/
 | bevy/
  | .cargo/
   | config.toml
   | config_fast_builds.toml
```

cart replied:
I'm starting to think that we just shouldn't use aliases. `cargo run --package ci` isn't that much harder to type than `cargo ci`. It is also clearer that it is a "package" in the workspace being run and not a "custom cargo command". And most importantly, it doesn't force us to stomp on config.toml (which is "user config").

CleanCut replied:
I'll make the change 👍🏼

CleanCut replied:
✅ `.cargo/config.toml` removed (removing the `cargo ci` alias, use `cargo run -p ci` or `cargo run --package ci` instead)

CleanCut commented:
If trailing newlines are important, we should add a check to CI. 

> I'd seen that but hadn't read it fully - in that case I'd argue we just shouldn't have a `rustfmt.toml` checked in at the root, and instead have it somewhere else. Then `cargo xtask fmt` would be `cargo +nightly fmt -- config-path <path_to_rusfmt_toml>`

AFAIK this would break @cart's desired experience. He uses nightly, and wants those options to take affect in his environment.

DJMcNab commented:
> this would break @cart's desired experience. 

I would suggest that in that case he should configure his format on save to be pointed to the new location of rustfmt.toml. When he (anyone) runs the catchup format they would just run `cargo xtask fmt` to apply it to the whole workspace.

CleanCut commented:
> > this would break @cart's desired experience.
> 
> I would suggest that in that case he should configure his format on save to be pointed to the new location of rustfmt.toml. When he (anyone) runs the catchup format they would just run `cargo xtask fmt` to apply it to the whole workspace.

I'm totally fine doing it whichever way @cart wants. Just let me know!

cart commented:
My vote is that we move away from aliases entirely in favor of `cargo run --package ci`. Its slightly more typing, but it allows us to not stomp on user config and is more self-documenting.

cart commented:
And (in a separate pr) I think we should add `cargo run --package fmt`, which runs stable `cargo fmt --all`.

Users (like me) that want to use nightly features can just set nightly as default, uncomment the relevant nightly config items, and run `cargo run --package fmt`

cart commented:
Or (alternatively) users can just run `cargo fmt --all``. Which is much simpler / not custom :)

CleanCut commented:
@cart Other than doing it in a separate PR, I'm not sure what you want done with `fmt`.  Currently we have `cargo fmt --all -- --check` as one of the things that runs inside of the `ci` package.  Do you want that to remain there?  Do you want it changed / removed?

cart commented:
Yeah sorry I wasn't super clear. I think we should leave `ci` as it is (check only). I don't think we need any more packages. Users can use the ci package to validate and `cargo fmt --all` to format.

CleanCut commented:
@cart Ready for another review!  I updated the issue to reflect the approach we ended up taking. I'll repeat it here for convenience:

> The final approach we settled on was to comment out the unstable options in `rustfmt.toml`. Those who are using `nightly` may  uncomment the unstable options locally if they wish. Once the options stabilize, we can uncomment them again.
> 
> We also decided that instead of fixing the alias, we would remove the alias entirely so that we do not introduce a custom `.cargo/config.toml` that would conflict with users' custom version of the same file. This means that instead of using a `cargo ci` alias you should use `cargo run -p ci` or `cargo run --package ci` instead.

I also updated the documentation on the website in https://github.com/bevyengine/bevy-website/pull/106

cart commented:
Ooh i didn't know `cargo run -p ci` was an option. Lets just recommend that everywhere instead of including both. I'm happy making the "terseness" decision on behalf of users.

Great work here. I think we found an ideal path forward :)

CleanCut commented:
> Great work here. I think we found an ideal path forward :)

I think everyone will be happy with it. 😄
